### ISSUE TITLE
method invocation syntax

### ISSUE THREAD

**Issue by dvv** [leafo/moonscript#14]
Hi!

Imho, back-slash style of method invocation is not eye pleasant. Back-slashes mean escaping of special characters in most languages.

Wonder if we could allow `instance:method(...)` syntax instead? We have such double meaning for `-`, which when used immediately before a token means negation, otherwise subtraction. Likewise, MoonScript's `instance:method(...)` could compile to Lua's `instance:method(...)`, while `instance: method(...)` could compile like at present, to `{instance = method(...)}`.

What do you think?

TIA,
--Vladimir

**Comment by nanoant** [comment]
Honestly I don't understand why we just cannot use `:` for method invocation like in Lua.

I understand this can collide with `:something` which means `{ something = something }`, but IMHO method call is much widely used in code than the above, so it should have kind of syntax priority.

So hereby I propose better non-abigious syntax for `{ something = something }`:

```moonscript
moon = 1
sun = 2
hash = moon:, sun:
```

* Simply not-specyfying value for hash, means value of the variable with same name as field. Simple as that!

Then we can switch back to `:` for method calls! I know that such change would break existing code-base, but... it will improve readability especially for those who come from C++ / Lua.
Otherwise current `\` syntax is really showstopper for many ppl.

**Comment by nilnor** [comment]
@nanoant I personally agree that `:` would have been preferable, but it seems unlikely to change unfortunately. You can have a look at issue #35 which also discusses this in more length.

**Comment by nanoant** [comment]
@nilnor Well, then it is highly possible that due this weird syntax this language will simply not gain more interest than it could. I think @leafo needs to acknowledge this is really an issue.

**Comment by nilnor** [comment]
@nanoant I agree, I would very much like to hear from @leafo again on this, even if it's only to say that it won't be changed.

**Comment by leafo** [comment]
No plans to change it. What's the problem you're having with `\`?

**Comment by nilnor** [comment]
I've expounded on this in issue #35, but at the risk of repeating myself I'll try to summarize:

In most other languages the backslash is used for escape sequences. Presumably it's this that makes the `\` feel so unnatural, but at any rate it sticks out like an eyesore. It will soon be three years after I started writing moonscript, and the `\` is not something that I've grown used to over time, so at least for me it's not a question of initial reluctance.

While it's anecdotal evidence, the comments in this issue and issue #35 seems to indicate that I'm not alone in considering it an issue. Personally, introducing moonscript to colleagues I have uniformly received the same feedback - they like it well enough but the backslash thing is a real turnoff making them not want to use it. It's a barrier to adoption.

And while perhaps a weaker argument, the `\` key is easy to type on a US layout keyboard but that does not hold true for other international keyboard layouts where it requires various degrees of RSI-inducing finger gymnastics.

In short, it's the one big drawback with moonscript in my opinion, and it's a shame considering the rest is so good.

**Comment by leafo** [comment]
Sorry, that question was directed to @nanoant , I already know your thoughts.

Honestly the only argument that I agree with is that it's difficult to type on non-us keyboards. That was something I definitely overlooked when picking it.

Any complaints about how it appears I just can't agree with. I don't know how to respond. I written however much MoonScript over the past X years, I see `\`, I know what it means, I don't even think about it. It never offends whatever stylistic sensibilities I've developed as a programmer. It's completely invisible in the same sense that I don't sit there and ponder why any other character is used for whatever it is used for.

Maybe it's analogous to 1 indexed arrays in Lua. People will always complain about it because it's different to what they are used to.

Okay, so then lets say I **fix** this. Do I replace it? Break all existing MoonScript that's been written? (There is quite a bit). Okay, write a tool that converts it? What about the thousands of lines of documentation and examples? Write more tools? Great, I'd love to waste all my time. If I want to really hurt adoption then I would could do something like that.

Alright, then what about adding alternate syntax? What is it going to be? Two characters? There really aren't any options for single characters that wouldn't be equally as offensive. Okay, so then we have more inefficient way of typing the same thing. That's effectively me failing as a language designer. Then all documentation and examples become confusing to beginners? No consistency? Am I supposed to phase out the `\`? I like using it.

Does `\` actually hurt adoption? Are the people who are upset over it going to start using MoonScript if it no longer has `\`. Does anyone actually have that thought process? "wow, this language used to use `\` but no more. I will now start using it." That just sounds ridiculous. I think it has a low adoption rate because it's already targeting such a niche group of programmers. I'm not disappointed about that, it's been an incredibly useful tool for me.

**Comment by stevedonovan** [comment]
On Thu, Feb 19, 2015 at 11:40 AM, leaf <notifications@github.com> wrote:

> Alright, then what about adding alternate syntax? What is it going to be?
> Two characters? There really aren't any options for single characters that
> wouldn't be equally as offensive.
>


Heh, we could always map '->' onto '\'.  ;)

But I agree, the language is what it is.  The example of 1-based array
indexing is good, because it remains a sticking point based on a feeling.
I don't think it scares away too many people, once they have made a
pragmatic decision to use Lua based on its merits.

**Comment by nilnor** [comment]
> Maybe it's analogous to 1 indexed arrays in Lua. People will always complain
about it because it's different from what they are used to.

Perhaps, and I guess that will come down to personal preferences. 1-indexed arrays was awkvard to me at the start as well, but that faded away quickly in a way the backslash hasn't.

> Does `\` actually hurt adoption?  \[..\] That just sounds ridiculous. 

I don't think anyone believes that revisiting the backslash would propel moonscript into a mainstream language. Lua itself is can be seen as niche language, and Moonscript will logically have an even smaller user base. But when people tell me they rather stay with Lua and cite the backslash as a factor in their decision I don't know how else to frame it.

I realize that changing this would be disruptive, that you see good reasons for not changing it, and I take it from your comments that that's the end of it. And that's fine - I appreciate you taking the time to make it clear.

I still believe that the `:` could be adjusted to cover this in an acceptable way, but I'll experiment on my own when time allows.

**Comment by nanoant** [comment]
> Sorry, that question was directed to @nanoant , I already know your thoughts. (...)
> Any complaints about how it appears I just can't agree with. I don't know how to respond. I've written however much MoonScript over the past X years. I see \ and I know what it means, (...).

My problem is I come from C++ & Ruby background, I feel comfortable with Lua. MoonScript seems to be great Lua meta syntax sugar language, with good borrows from Ruby. But honestly I don't understand why I need to learn myself (teach my brain & eyes) 3rd method call scheme when reading the code!? I believe there is absolutely no justification for that, no reason for the change.

I wish MoonScript addressed only Lua's shortcoming, but not change something that worked well. I am and I will be working in language mixed environment and it really stings my eyes when I read MoonScript sources.

Also I believe you and existing MoonScript users will not suffer as much when adapting `:` way back since anyway you code in Lua and this will be consistent.

3rd and maybe last important argument is that method signatures in docs will be the same is we keep `:` and `.` to have same meaning across MoonScript and Lua.

> (...) But when people tell me they rather stay with Lua and cite the backslash as a factor in their decision I don't know how else to frame it.

This couldn't be expressed better. This issue and previous one come with some reason. Indeed it isn't really about making MoonScript mainstream language, but rather making MoonScript a CoffeeScript for Lua and ease its adoption for existing Lua users.

**Comment by tophf** [comment]
I wonder how many moonscript users besides the language author would advocate so ardently for keeping `\`? So far I'd say there won't be any... especially in case the new version supports both `\` and e.g. `->`, issuing a compilation warning, probably.

**Comment by kirbyfan64** [comment]
Ok...I'm not a very common MoonScript user...but...

While I *do* agree that `\` is slightly confusing, I don't think `:` should be adapted, because that is going to break 75% of all MoonScript code ever written. It'd be even more confusing (`{a:b} != {a: b}`, which is bound to lead to confusing bugs all over code).

I think `->` would be overly C-esque...but it could probably be used, since function definition is a prefix operator (I think), and it lies well with the C++ concept of classes (I'm assuming that's a good thing :).

I personally think that `\` is easy to catch on to, though, and it's very unambiguous once you get used to reading it.

Overall...+1 for leaving it as-is, but, if it *needs* to change, let it be `->`, not a weird overload of `:`.

**Comment by nanoant** [comment]
> I think -> would be overly C-esque...but it could probably be used, since function definition is a prefix operator (I think), and it lies well with the C++ concept of classes (I'm assuming that's a good thing :).

But in C it has opposite meaning `->` of dereference pointer and access field. Also in Objective-C it means direct struct field access, where `.` means dynamic property call. Also it will visually collide with lambdas, also there will be some ambiguity here:
```moonscript
(str)->str -- I am calling of defining function here!?
``` 

Honestly I give up, I cannot think only sensible alternative. Last chance I see is a tick character \` it doesn't glue so much with surrounding, provides visual separation between callee and method name:

```moonscript
class Test
  method: => 1
  static: -> "else"

-- old backslash
Test\method!

-- tick notation
Test`method!

-- static method call
Test.static!
```
But it is same (un)easy to type as backslash.

**Comment by tophf** [comment]
Well then the doublecolon `::` suggestion by dvv above seems like an obvious winner but I guess someone will have to fork Moonscript for that because the author definitely won't condescend to us the weaklings that stumble over such a triviality.

**Comment by kirbyfan64** [comment]
I don't like double-colon either; I just prefer the backslash. I'm used to it.

**Comment by szensk** [comment]
@tophf It's backwards compatibility. It's a mainly aesthetic change. For some bizarre reason `\` does upset my sensibilities but it was already chosen. To do busy work to add a `->` or `::` as an alternate syntax for `\` is really a waste of time. Something that is best left to forks, really.

Honestly, to change it to `::` required 5 replaces of `//` in parse.moon. I'm not sure if it is ideal, but if you're distributing lua code instead of moonscript, you can just roll your own.
